content/learning-paths/servers-and-cloud-computing/exploiting-stack-buffer-overflow-aarch64/frame-layout.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Frame Layout
+weight: 4
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+To be able to build a stack buffer overflow-based exploit, we first need to
+understand how variables, function arguments, return values and more are laid
+out on the stack.
+
+In this section, we will look at how the `clang` compiler lays out these
+entities.
+
+Let's look at the following program:
+
+```C
+__attribute__((noinline))
+long f() {
+    volatile char buffer1[4];
+    volatile char buffer2[7];
+    buffer1[3] = 10;
+    buffer2[6] = 20;
+    return buffer1[3]+buffer2[6];
+}
+
+int main() {
+    f();
+    return 0;
+}
+```
+Save it as file `stack-layout.c`.
+
+Compile it with the following command at your docker prompt:
+
+```bash { command_line="root@7a8fb34f810e:/armlearningpaths|2-3" }
+clang -O1 -S -fno-asynchronous-unwind-tables stack-layout.c
+```
+
+This produces the assembly file `stack-layout.s`. When you open it, you see the
+following assembly for function `f`:
+
+```asm
+	sub	sp, sp, #16
+	mov	w8, #10
+	mov	w9, #20
+	strb	w8, [sp, #12]
+	strb	w9, [sp, #8]
+	ldrb	w8, [sp, #12]
+	ldrb	w9, [sp, #8]
+	add	x0, x9, x8
+	add	sp, sp, #16
+	ret
+```
+
+Here we see that the first instruction `sub sp, sp #16` reserves 16 bytes on the
+stack as the stack frame for this function. The stack on AArch64 grows
+downwards, i.e. towards lower memory addresses.
+
+The next 4 instructions store value `10` and `20` to `buffer1[3]` and
+`buffer2[6]`.
+
+From the offset `[sp, #8]` and `[sp, #12]` we can compute the frame layout.
+
+{{% notice "exercise 1" %}}
+
+Draw the frame layout of function `f`: how are `buffer1` and `buffer2` laid out
+in the 16 bytes that are reserved?
+
+TODO: find a way to hide the answer, e.g. by some form of collapsible UI element?
```

Comment:
I still need to find a nice way for the answer to an exercise to not be immediately visible.
I guess possibilities would be:
1. Implement some form of "collapsible" UI where the answer details are collapsed, and the reader that wants to see the answer "uncollapses" it.
2. Always split a section at an exercise, and have the answer in a separate section. That way the reader has to actively click "next" to see the answer, rather then it just being there on the screen.

Reply:
I decided to put all answers in a separate section and provide a hyperlink from the exercise to the answer. See commit 9f50876